### PR TITLE
NEW: Mock sleep unit test utility.

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# NEW: Mock sleep unit test utility.

Simple utility function for unit tests. Use instead of `sleep()`. I found this quite handy is several projects, we might as well make the code reusable.

### CODE REVIEW

**Comment by GuySartorelli** [comment]
I don't think we should be including methods that are only meant to be called in tests into production codebases.
There is some of that already, but it's too late to decouple those instances in the current major. I don't think we should be adding more.

I'll leave this open for a day or two in case a core committer disagrees with me, though.

**Comment by mfendeksilverstripe** [comment]
@GuySartorelli Fair enough, is there a better place for it though? For example we could add a TestOnly service maybe under the Dev namespace?

**Comment by GuySartorelli** [comment]
I'm not sure, to be honest. I'd love (in the long term) for all of the test-related stuff to be moved out of framework altogether, and be in its own module that can be included as a dev dependency... but we're a ways off doing something like that just yet.

This might be the best place for it at the moment, to be honest, now that I've spent a bit more time considering it.
I still won't merge immediately in case there are opinions from other core committers on this but yeah probably this is the spot for this method, if there is a spot for it, in the current state of things.

**Comment by mfendeksilverstripe** [comment]
All good @GuySartorelli , happy to keep the PR open. Having a separate module for this sounds great.

**Review by michalkleiner — APPROVED** [review]
Fair consideration from Guy, I would agree here. We can wait forever or just get it in now and move it out later with other stuff.

**Comment by michalkleiner** [comment]
Let's wait a day or two if Max or Steve have a different view, and if not we can get it in.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
This does seem like the wrong place

Somewhere in src/Dev seems like it would be better.  There are already some other Test*.php classes there such as TestSession and TestMailer

So perhaps create a new class Dev/TestDateTime.php and put this method it there?

**Comment by dhensby** [comment]
Couldn't it be added to `SapphireTest` or some part of the PHPUnit framework to allow tests to call `$this->addTime()` or similar?

How would one expect this to work if `mock_now` was not set?

**Comment by emteknetnz** [comment]
I'd be inclined to just delete the PHPUnit 5.7 version, it's an ancient API we shouldn't be developing for it

What happens if you put a `sleep(3)` in your unit test?

**Comment by GuySartorelli** [comment]
> I'd be inclined to just delete the PHPUnit 5.7 version,

Just for the sake of clarity: I assume you mean the 5.7 version _of the new method_, not of the class itself.

**Comment by dhensby** [comment]
Cool. Is the time reset after each test? I assume it is, just making sure...

**Comment by GuySartorelli** [comment]
If it isn't, that's a fundamental issue with the mock now functionality as a whole, not a new problem introduced with this PR.

**Comment by dhensby** [comment]
Right, but a dev who implements `mock_now` is then responsible for the tear down of that test. This feature hides it behind syntactic sugar, so it'd be understandable that a dev that uses it assumes the date is reset at the end of the test for them.

**Comment by dhensby** [comment]
Looks like it is handled: https://github.com/silverstripe/silverstripe-framework/blob/4/src/Dev/SapphireTest.php#L630-L632

**Comment by GuySartorelli** [comment]
@emteknetnz You've still got a review here with requested changes - is that for removing the new method on the old phpunit class? I don't think it's worth holding up merging for that personally, are you okay with the PR as is or do you still want some changes made?

**Comment by emteknetnz** [comment]
I'm not too attached, feel free to merge if you see fit

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**
